File: packages/var-core/src/doc-string-diff.ts
import { ReturnShapeError } from './cell-diff.js'
import type { Span } from './span.js'

// A doc-string content difference: the fence body's source range plus the
// expected (authored) and actual (returned) strings.
export type DocStringDiff = {
  readonly span: Span
  readonly expected: string
  readonly actual: string
}

// Compare a doc-string step's returned string against the fence body content.
// Exact equality (the body includes its trailing newline). `undefined` → no
// check (null). A non-string return is an author mistake → ReturnShapeError.
export function compareDocString(
  returned: unknown,
  content: string,
  span: Span,
): DocStringDiff | null {
  if (returned === undefined) return null
  if (typeof returned !== 'string') {
    throw new ReturnShapeError(`expected a doc string (string), got ${typeof returned}`)
  }
  if (returned === content) return null
  return { span, expected: content, actual: returned }
}

// Thrown by the executor when a doc-string step's returned string differs.
export class DocStringMismatchError extends Error {
  readonly diff: DocStringDiff
  constructor(diff: DocStringDiff) {
    super(
      `doc string: expected ${JSON.stringify(diff.expected)} but was ${JSON.stringify(diff.actual)}`,
    )
    this.name = 'DocStringMismatchError'
    this.diff = diff
  }
}

export function isDocStringMismatchError(e: unknown): e is DocStringMismatchError {
  return e instanceof DocStringMismatchError
}

File: rust/var-core/src/doc_string_diff.rs
//! Doc-string comparison — port of `doc-string-diff.ts` / `DocStringDiff.java`.

use crate::error::StepError;
use crate::span::Span;
use crate::value::Value;

/// A doc-string content difference: the fence body's span plus expected/actual.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct DocStringDiff {
    pub span: Span,
    pub expected: String,
    pub actual: String,
}

impl DocStringDiff {
    pub fn new(
        span: Span,
        expected: impl Into<String>,
        actual: impl Into<String>,
    ) -> DocStringDiff {
        DocStringDiff {
            span,
            expected: expected.into(),
            actual: actual.into(),
        }
    }
}

/// Compares a doc-string step's return against the fence body (exact equality,
/// trailing newline included). `None` → no check. A non-string return →
/// [`StepError::ReturnShape`].
pub fn compare_doc_string(
    returned: Option<&Value>,
    content: &str,
    span: Span,
) -> Result<Option<DocStringDiff>, StepError> {
    let s = match returned {
        None => return Ok(None),
        Some(Value::String(s)) => s,
        Some(other) => {
            return Err(StepError::ReturnShape(format!(
                "expected a doc string (string), got {}",
                other.type_name()
            )));
        }
    };
    if s == content {
        Ok(None)
    } else {
        Ok(Some(DocStringDiff::new(span, content, s.clone())))
    }
}
